[clang-tidy][dataflow] Add `bugprone-null-check-after-dereference` check

Discookie opened the pull request:
This check implements a null-pointer analysis using the data-flow framework. In particular, the check finds cases where the pointer's nullability is checked after it has been dereferenced:

```cpp
int f(int *ptr) {
  *ptr += 20;   // note: one of the locations where the pointer's value cannot be null
  
  if (ptr) {
    *ptr += 42; // warning: pointer value is checked even though it cannot be null at this point
    return *ptr;
  }
  return 0;
```

### It currently recognizes the following operations

- Dereference and member access through arrow operators
- Pointer-to-boolean cast
- Assigning `&obj`, `nullptr`, and other values

### Notable limitations

The checker only supports C++ due to the limitations of the framework (llvm/llvm-project#65301).

Function calls taking a reference of a pointer are not handled yet.
``void external(int *&ref);``

The note tag is only displayed at one location, and not all operations are supported or displayed.

More examples and limitations in the [checker docs](https://github.com/Discookie/llvm-project/blob/reverse-null/clang-tools-extra/docs/clang-tidy/checks/bugprone/null-check-after-dereference.rst).


### Results on open-source projects

Available at: https://codechecker-demo.eastus.cloudapp.azure.com/Default/runs?run=Discookie_reverse-null

The reference-of-ptr limitation leads to a class of false positives across the tested projects [(example)](https://codechecker-demo.eastus.cloudapp.azure.com/Default/report-detail?run=libwebm_libwebm-1.0.0.27_Discookie_reverse-null&detection-status=New&report-id=3497414&report-hash=e75b81fcd825616e0ec904ffef1c99ac&report-filepath=%2Flocal%2Fpersistent_docker%2FCSA-measurements-driver-2923%2Fmeasurements_workspace%2Flibwebm%2Fmkvparser.cpp).

But the checker also found quite a few true positives, on LLVM: [[1]](https://codechecker-demo.eastus.cloudapp.azure.com/Default/report-detail?report-id=3503712) [[2]](https://codechecker-demo.eastus.cloudapp.azure.com/Default/report-detail?report-id=3503771) [[3]](https://codechecker-demo.eastus.cloudapp.azure.com/Default/report-detail?report-id=3503770) and a couple more, listed [here](https://github.com/Discookie/llvm-project/pull/1#issuecomment-1979214081).


---

This checker corresponds to the the CERT rule [EXP34-C. Do not dereference null pointers, example #3](https://wiki.sei.cmu.edu/confluence/display/c/EXP34-C.+Do+not+dereference+null+pointers#EXP34C.Donotdereferencenullpointers-NoncompliantCodeExample.2).

whisperity reviewed (commented):
(@Discookie and I had an extensive prior review at Discookie/llvm-project#1 in case someone is interested in the details.)

Discookie committed:
[clang][dataflow] Add null-check after dereference checker

EugeneZelenko commented on clang-tools-extra/docs/clang-tidy/checks/bugprone/null-check-after-dereference.rst:
```diff
@@ -0,0 +1,162 @@
+
+   This check uses a flow-sensitive static analysis to produce its
+   results. Therefore, it may be more resource intensive (RAM, CPU) than the
+   average clang-tidy check.
```
`Clang-tidy`.

EugeneZelenko commented on clang-tools-extra/docs/clang-tidy/checks/bugprone/null-check-after-dereference.rst:
```diff
@@ -0,0 +1,162 @@
+
+The checker currently supports null-checks on pointers that use
+``operator bool``, such as when being used as the condition
+for an `if` statement.
```
Please use double back-ticks for language constructs.

EugeneZelenko commented on clang-tools-extra/docs/clang-tidy/checks/bugprone/null-check-after-dereference.rst:
```diff
@@ -0,0 +1,162 @@
+Null-pointer and other value assignments
+----------------------------------------
+
+The checker supports assigning various values to pointers, making them *null*
```
`check`. Same in other places.

EugeneZelenko commented on clang/lib/Analysis/FlowSensitive/Models/NullPointerAnalysisModel.cpp:
```diff
@@ -0,0 +1,625 @@
+}
+
+Value *getValue(const Expr &Var, Environment &Env) {
+  if (auto *EnvVal = Env.getValue(Var)) {
```
Please do not use `auto` if type is not explicitly stated in same statement or iterator. Same in other places.

EugeneZelenko commented on clang/lib/Analysis/FlowSensitive/Models/NullPointerAnalysisModel.cpp:
```diff
@@ -0,0 +1,625 @@
+void initializeRootValue(Value &RootValue, Environment &Env) {
+  Arena &A = Env.arena();
+
+  BoolValue *IsNull = cast_or_null<BoolValue>(RootValue.getProperty(kIsNull));
```
`auto` could be used, because type is stated in same statement.

EugeneZelenko reviewed (commented):
Please mention new check in Release Notes.

gribozavr commented on clang/lib/Analysis/FlowSensitive/Models/NullPointerAnalysisModel.cpp:
```diff
@@ -0,0 +1,625 @@
+
+auto castExprMatcher() {
+  return castExpr(hasCastKind(CK_PointerToBoolean),
+                  hasSourceExpression(ptrToVar()))
```
Why limit the cast argument to a var? It shouldn't matter if the argument is a complex expression. As long as the framework models the value somehow (for example, one of the implicit casts that the core framework already models, like the lvalue to rvalue cast), the transfer function should not need to worry about how that ended up happening.

martinboehme replied:
> I was afraid that a more complex argument would run into issues with being rvalue vs lvalue, and the differences between the two had me crash the framework a couple times before.

If this is causing issues for you and you're not clear on why, let me know -- happy to go over the details of how this works.

Nit: The relevant distinction is actually between:

*  prvalue -- a "pure rvalue" that is "just a value" (it has no storage location / address, i.e. it's not permissible to apply the address-of operator `&` to it). This is represented in the framework as a `Value` (or some subclass)

*  glvalue -- a "generalized lvalue", i.e. something that has a storage location / address, which you can obtain using the `&` operator. This is represented in the framework as a `StorageLocation` (or `RecordStorageLocation`).

Discookie replied:
Indeed they are glvalues/prvalues.
The issue I encountered earlier was that there are a couple unintuitive operators, such as prefix `++` as glvalue, that can appear in unexpected places, where there should only be a prvalue otherwise. Which leaves the model to decide which `setValue` to call.
Looks like it's not an issue for now as I haven't found any new crashes.

Discookie replied:
No longer limited in the general case either - works much better that way indeed. Way fewer value-type mismatches as well.

gribozavr commented on clang/lib/Analysis/FlowSensitive/Models/NullPointerAnalysisModel.cpp:
```diff
@@ -0,0 +1,625 @@
+
+    Env.assume(A.makeNot(IsNonnull.formula()));
+    NewRootValue->setProperty(kIsNonnull, Env.getBoolLiteralValue(false));
+  }
```
I think this function could be significantly simplified.

You should check if the Expr argument of the cast has a modeled pointer value, and if it is, then one of the booleans that you're storing as synthetic properties on that pointer value would be the result of the pointer to bool cast.

See https://github.com/google/crubit/blob/085108ea0a3b8da2e400613d7d235274de1f34b9/nullability/pointer_nullability_analysis.cc#L822 for a sample transfer function for a cast.

If not all variables reliably get a modeled pointer value (I'm guessing this is the case because a good chunk of this transfer function is dedicated to creating values), then that problem should be solved uniformly for all cases. PTAL at `unpackPointerValue`, `initPointerNullState` and `initPointerFromTypeNullability` in our code, and how they are used - specifically `transferValue_Pointer` and that it is called for every pointer-typed expression, not just variables. For the initialization of synthetic properties (`kIsNull` etc.) the idea is to call initialization in a central place, so that each individual transfer function does not need to worry about it.

`unpackPointerValue` also handles "top" and I think your code would benefit from something similar, as it leads to significantly better convergence of loop analysis.

Discookie replied:
For now I am only doing what you said about `transferValue_Pointer`, and only matching special cases where a pointer is used instead of created. I'm in the process of simplifying all of the logic along the lines of this, though.

gribozavr commented on clang/lib/Analysis/FlowSensitive/Models/NullPointerAnalysisModel.cpp:
```diff
@@ -0,0 +1,625 @@
+
+auto functionCallMatcher() {
+  return callExpr(hasDeclaration(functionDecl(returns(isAnyPointer()))))
+      .bind(kVar);
```
Please see my other comment about following the `transferValue_Pointer` pattern from our nullability checker - if you do that you wouldn't need to define multiple matchers to cover all C++ syntax that could produce a pointer. Instead it would be only one matcher that covers all pointer-valued expressions, which should be simpler.

martinboehme commented on clang-tools-extra/clang-tidy/bugprone/NullCheckAfterDereferenceCheck.cpp:
```diff
@@ -0,0 +1,171 @@
+  NullCheckAfterDereferenceDiagnoser Diagnoser;
+  NullCheckAfterDereferenceDiagnoser::ResultType Diagnostics;
+
+  using LatticeState = DataflowAnalysisState<NullPointerAnalysisModel::Lattice>;
```
Nit: This isn't just the "lattice state" -- it's the entire state, comprising the `Environment` and the lattice. Maybe simply `State`?

Discookie replied:
`State` works, yep.

martinboehme commented on clang-tools-extra/test/clang-tidy/checkers/bugprone/null-check-after-dereference.cpp:
```diff
@@ -0,0 +1,330 @@
+    // CHECK-MESSAGES: :[[@LINE-4]]:3: note: one of the locations where the pointer's value cannot be null
+  // FIXME: If there's a direct path, make the error message more precise, ie. remove `one of the locations`
+    *p += 20;
+    return *p;
```
Nit: This line seems to be redundant -- we already have a `*p` in the same block. Also, why does this function need to return anything -- just make it `void`?

(Generally, try to make tests contain only the minimal amount of code you need to test the behavior the test targets. Also applies similarly to other tests below.)

Discookie replied:
Removed a couple of the redundant lines, but I do wanna go over the tests once again, and rewrite them to be more concise.

martinboehme commented on clang-tools-extra/test/clang-tidy/checkers/bugprone/null-check-after-dereference.cpp:
```diff
@@ -0,0 +1,330 @@
+  }
+
+  if (p) {
+    // CHECK-MESSAGES-NOT: :[[@LINE-1]]:7: warning: pointer value is checked even though it cannot be null at this point 
```
I think this is redundant. This style of test will fail if there are any unexpected diagnostics.

martinboehme commented on clang-tools-extra/test/clang-tidy/checkers/bugprone/null-check-after-dereference.cpp:
```diff
@@ -0,0 +1,330 @@
+  }
+}
+
+int two_branches_reversed(int *p, bool b) {
```
This test seems borderline redundant -- it's really testing the framework itself rather than your check.

martinboehme commented on clang-tools-extra/test/clang-tidy/checkers/bugprone/null-check-after-dereference.cpp:
```diff
@@ -0,0 +1,330 @@
+  }
+
+  if (nullptr_assigned) {
+    // CHECK-MESSAGES-NOT: :[[@LINE-1]]:7: warning: pointer value is checked even though it cannot be null at this point
```
Redundant

martinboehme commented on clang-tools-extra/test/clang-tidy/checkers/bugprone/null-check-after-dereference.cpp:
```diff
@@ -0,0 +1,330 @@
+  }
+}
+
+extern int *fncall();
```
```suggestion
extern int *returns_int_ptr();
```

Naming nit: The function itself isn't a "function call".

martinboehme commented on clang-tools-extra/test/clang-tidy/checkers/bugprone/null-check-after-dereference.cpp:
```diff
@@ -0,0 +1,330 @@
+  }
+}
+
+int chained_references(int *a, int *b, int *c, int *d, int *e) {
```
Does this really need a chain this long?

martinboehme replied:
Can you remove the unused parameters `d` and `e` too?

martinboehme commented on clang/include/clang/Analysis/FlowSensitive/Models/NullPointerAnalysisModel.h:
```diff
@@ -0,0 +1,112 @@
+public:
+  /// A transparent wrapper around the function arguments of transferBranch().
+  /// Does not outlive the call to transferBranch().
+  struct TransferArgs {
```
Why does this need to be public? It doesn't seem to be used outside `NullPointerAnalysisModel`?

Discookie replied:
Deleted.

martinboehme commented on clang/include/clang/Analysis/FlowSensitive/Models/NullPointerAnalysisModel.h:
```diff
@@ -0,0 +1,112 @@
+  // Used to initialize the storage locations of function arguments.
+  // Required to merge these values within the environment.
+  void initializeFunctionParameters(const ControlFlowContext &CFCtx,
+                                    Environment &Env);
```
It this used outside this class? (It looks as if it could be private?)

martinboehme commented on clang/include/clang/Analysis/FlowSensitive/Models/NullPointerAnalysisModel.h:
```diff
@@ -0,0 +1,112 @@
+  llvm::DenseMap<const Value *, const Expr *> ValToDerefLoc;
+  // Maps Maps a warning's SourceLocation to its relevant Value.
+  llvm::DenseMap<SourceLocation, const Value *> WarningLocToVal;
+
```
Do these need to be public? They look like internal implementation details of the class.

martinboehme replied:
It looks, though, as if these maps are passed to the diagnose-match-switch through a `DiagnoseArgs` parameter, rather than being accessed directly. So I think these could (and should) still be private?

martinboehme commented on clang/lib/Analysis/FlowSensitive/Models/NullPointerAnalysisModel.cpp:
```diff
@@ -0,0 +1,625 @@
+  const auto MergeValues = [&](llvm::StringRef Name) -> BoolValue & {
+    BoolValue *LHSVar = cast_or_null<BoolValue>(Val1.getProperty(Name));
+    BoolValue *RHSVar = cast_or_null<BoolValue>(Val2.getProperty(Name));
+
```
I think you want an early-out here along the lines of:

```cxx
if (LHSVar == RHSVar)
  return *LHSVar;
```

martinboehme commented on clang/lib/Analysis/FlowSensitive/Models/NullPointerAnalysisModel.cpp:
```diff
@@ -0,0 +1,625 @@
+  auto CompareValues = [&](llvm::StringRef Name) -> ComparisonResult {
+    BoolValue *LHSVar = cast_or_null<BoolValue>(Val1.getProperty(Name));
+    BoolValue *RHSVar = cast_or_null<BoolValue>(Val2.getProperty(Name));
+
```
Again, you'll want an early-out for `LHSVar == RHSVar`.

martinboehme commented on clang/lib/Analysis/FlowSensitive/Models/NullPointerAnalysisModel.cpp:
```diff
@@ -0,0 +1,625 @@
+    BoolValue *LHSVar = cast_or_null<BoolValue>(Val1.getProperty(Name));
+    BoolValue *RHSVar = cast_or_null<BoolValue>(Val2.getProperty(Name));
+
+    SatisfiabilityResult LHSResult = computeSatisfiability(LHSVar, Env1);
+    SatisfiabilityResult RHSResult = computeSatisfiability(RHSVar, Env2);
```
This looks like a pretty expensive way of evaluating a comparison: You're potentially calling the SAT solver four times (twice for each `computeSatisfiability()`).

Instead, I'd suggest testing "proves(makeEquals(*LHSVar, *RHSVar))" and "proves(makeNot(makeEquals(*LHSVar, *RHSVar)))" (pseudocode), which gives you at most two SAT solver invocations.

Discookie replied:
The issue with evaluating something similar to `LHS == RHS` is that in the environments `Env1`, `Env2` and `MergedEnv` can have different satisfiability results.
I can imagine evaluating something similar to `(FlowCondition1 => LHS) == (FlowCondition2 => RHS)` against `MergedEnv` could work, but I'll need to test that first.

martinboehme commented on clang/lib/Analysis/FlowSensitive/Models/NullPointerAnalysisModel.cpp:
```diff
@@ -0,0 +1,625 @@
+
+// Computes satisfiability by using the flow condition. Slower, but more
+// precise.
+SatisfiabilityResult computeSatisfiability(BoolValue *Val,
```
Naming nit: This isn't really computing satisfiability; it's computing whether or not the environment proves that a `BoolValue` is necessarily true or false.

I think you could rename this `tryToProve()`, and you could probably do without the `SatisfiabilityResult` type entirely (which isn't really an accurate name either) and instead return a `BoolValue &` (since that's what the caller ends up needing anyway). This would be either a literal true or false (if the environment can prove the value true or false) or `*Val` if the environment can't prove anything about the variable.

Discookie replied:
Returning a `BoolValue &` sounds useful. I wouldn't be able to differentiate cases such as `nullptr` vs `unknown`, but maybe I don't even need to, since I'm gonna initialize the variable later anyways.

martinboehme commented on clang/lib/Analysis/FlowSensitive/Models/NullPointerAnalysisModel.cpp:
```diff
@@ -0,0 +1,625 @@
+}
+
+void setUnknownValue(const Expr &E, Value &Val, Environment &Env) {
+  if (E.isGLValue())
+    setGLValue(E, Val, Env);
+  else if (E.isPRValue())
+    Env.setValue(E, Val);
+  else
+    llvm_unreachable("all value cases covered");
```
```suggestion
  if (E.isGLValue())
    setGLValue(E, Val, Env);
  else
    Env.setValue(E, Val);
```

`isGLValue()` <=> `!isPRValue()`

Discookie replied:
Indeed, fixed.

martinboehme commented on clang/lib/Analysis/FlowSensitive/Models/NullPointerAnalysisModel.cpp:
```diff
@@ -0,0 +1,625 @@
+}
+
+// Assigns initial pointer null- and nonnull-values to a given Value.
+void initializeRootValue(Value &RootValue, Environment &Env) {
```
Naming nit: Why "root" value? `initializePointerProperties()` would seem more descriptive to me here.

Discookie replied:
Indeed, renamed.

martinboehme commented on clang/lib/Analysis/FlowSensitive/Models/NullPointerAnalysisModel.cpp:
```diff
@@ -0,0 +1,625 @@
+    return;
+  }
+
+  auto *NewRootValue = Env.createValue(Var->getType());
```
`createValue()` is guaranteed to return nonnull for pointer types, so you can simply assert that the returned value is nonnull here.

Discookie replied:
That's good to know! Even if the contained type cannot be represented, it still returns the PointerValue normally. Fixed.

PiotrZSL reviewed (commented):
I like idea behind this check, and I think that there should be version of this check not only for raw pointers but also for optionals, smart pointers and iterators.

My main problem is that dataflow framework is slow and unstable, there are 20 issues open for an bugprone-unchecked-optional-access check that uses this framework and 19 issues for a framework alone. It crashes, it hangs and only cause problems.

Personally I would prefer check like this to be written in simpler way by using same method as bugprone-use-after-move is using. Even if it would find less issues, but at-least wouldn't force half of the projects to disable it due to crashes or long execution time.

Otherwise please consider implementing this check as an clang-analyzer check, so that all incoming issues could be redirected there instead pin to clang-tidy.

ymand commented:
> My main problem is that dataflow framework is slow and unstable, there are 20 issues open for an bugprone-unchecked-optional-access check that uses this framework and 19 issues for a framework alone. It crashes, it hangs and only cause problems.

I think those issues are stale. We've done a huge amount of stability engineering in the past 6 months and are running multiple checks based on this framework ~daily over our entire (> 500 MLoc) codebase with no known crashes.

The same goes for speed - it's definitely more expensive than typical, local, clang tidies, but we've made significant improvements in this area as well, and track performance of our nullability check with benchmark tests.

More generally, I think that if we push folks away from experimental clang tidy checks because of stability, we'll shut down experimentation for clang tidy. Is there no way to mark new/experimental checks so that they are off by default?

PiotrZSL commented:
> I think those issues are stale.
Last time I tested it it were on Clang-tidy 17. I'm doing migration to Clang-tidy 18 now. I will check this and bugprone-unchecked-optional-access check on my code-base, on which it were unstable previously. If nothing will hang/crash, then I will be more than happy.

ymand commented:
> > I think those issues are stale.
> 
> Last time I tested it it were on Clang-tidy 17. I'm doing migration to Clang-tidy 18 now. I will check this and bugprone-unchecked-optional-access check on my code-base, on which it were unstable previously. If nothing will hang/crash, then I will be more than happy.

Thanks! Note that one open issue is being fixed in https://github.com/llvm/llvm-project/pull/84138 (which I just realized has not landed yet as a commit). So, if you still see crashes, please let us know if that PR fixes it.

whisperity commented:
> Is there no way to mark new/experimental checks so that they are off by default?

@ymand At some point in the past, albeit several years ago, I had worked on a check (not with the data-flow framework!) which was originally on track to be introduced as such an *experimental* check. See https://reviews.llvm.org/D76545. The idea was that similarly to `alpha.` checkers in **CSA** we could add an `experimental-`. The biggest counter-argument is that there is no good policy for when something can be put in as experimental, and we don't have a consensus as to how long something can stay experimental, how experimental things are dropped, or promoted into non-experimental checkers, etc. **CSA** has several alpha checkers (some have been there for almost a decade now!), and some of my colleagues are working hard on improving them.

As of some 18.0 version, Clang-Tidy by default enables **no** checkers apart from those "inherited" from **CSA**. At least calling a raw `clang-tidy --list-checks` only gives back `clang-analyzer-` ones...

@PiotrZSL For interactive users, we have added the new check to the list of forbidden checks that are auto-enabled through clangd. Just like how the optional checker is on that list.

Discookie committed:
Reverse-null surface-level fixes

Discookie committed:
Fix crash when analyzing anonymous lambda fns

Discookie reviewed (commented):
Thanks for the reviews! Fixed a lot of the comments, and I have a few more things to test for the rest.

@gribozavr Thank you for the thorough introduction to your approaches, and the links to related works as well! Templated types are something I didn't take a look at before, will need to experiment with those as well.

EugeneZelenko commented on clang-tools-extra/docs/ReleaseNotes.rst:
```diff
@@ -116,6 +116,12 @@ New checks
   Replaces certain conditional statements with equivalent calls to
   ``std::min`` or ``std::max``.
 
+- New :doc:`bugprone-null-check-after-dereference
```
Please keep alphabetical order (by check name) in this section.

EugeneZelenko commented on clang-tools-extra/docs/clang-tidy/checks/bugprone/null-check-after-dereference.rst:
```diff
@@ -0,0 +1,162 @@
+   results. Therefore, it may be more resource intensive (RAM, CPU) than the
+   average Clang-tidy check.
+
+This check identifies redundant pointer null-checks, by finding cases where the
```
Please synchronize with statement in Release Notes.

martinboehme commented on clang-tools-extra/clang-tidy/bugprone/NullCheckAfterDereferenceCheck.cpp:
```diff
@@ -0,0 +1,171 @@
+  using namespace ast_matchers;
+
+  auto hasPointerValue =
+      hasDescendant(NullPointerAnalysisModel::ptrValueMatcher());
```
That link to the implementation doesn't work for me, and I still can't find the implementation by searching the diff. (When I asked where it's "defined", I specifically meant the implementation; the mention of the function in the header file is a declaration.)

martinboehme replied:
I've found the implementation now.

martinboehme commented:
Some general comments:

*  Is this ready for the next round of review? I see you've pushed a change, but you remark in other comments that there are still changes left to do. Can you let us know when this is ready for review again?

*  When updating the PR, can you [use fixups instead of force-pushing](https://llvm.org/docs/GitHub.html#updating-pull-requests)? This makes it easier to see what you've changed.

*  This PR has a _lot_ of reviewers on it. I'd suggest keeping the ones who have made comments so far and removing the others. (They can always re-add themselves if they want to, but a PR shouldn't generally need this many reviewers.)

martinboehme commented:
PS

*  Tests appear to be failing.

Discookie commented on clang/lib/Analysis/FlowSensitive/Models/NullPointerAnalysisModel.cpp:
```diff
@@ -0,0 +1,625 @@
+// FIXME: These AST matchers should also be exported via the
+// NullPointerAnalysisModel class, for tests
+auto ptrToVar(llvm::StringRef VarName = kVar) {
+  return traverse(TK_IgnoreUnlessSpelledInSource,
```
Apparently `TK_IgnoreUnlessSpelledInSource` is still required here. While getting `Value`s out of the framework is propagated correctly, *setting* values is not.
When setting a new `Value`, only the outermost `Expr` will have that value set, which is not guaranteed to be a glvalue even if the underlying value is, and could be overwritten by the propagation done by `Transfer.cpp`. To set a value directly, we still need to bind to the root `Expr` in this case.
This traversal has weird behavior with `nullptr` and some other casts - on `nullptr` we need to match the `NullToPointer` cast, and not the nullptr directly, and explicit casts also have the chance to not bind correctly to the root pointer.
I'm not sure if there's a better way to bind to the "root" `Expr` that I haven't found yet, but the current solution has too many dragons. (Maybe this is also a case of X/Y problem, where if I assign atoms as `IsNull/IsNonnull` instead of formulas, I wouldn't need to re-assign `Value`s...)

martinboehme replied:
> I'm not sure if there's a better way to bind to the "root" Expr that I haven't found yet, but the current solution has too many dragons. (Maybe this is also a case of X/Y problem, where if I assign atoms as IsNull/IsNonnull instead of formulas, I wouldn't need to re-assign Values...)

I think this is the underlying issue. When you glean new information about an existing `Value`, that information should be reflected in the flow condition, rather than by replacing the existing value.

As you've discovered, replacing the value of a prvalue expression with a different value does nothing to change the value of a glvalue expression that the prvalue may have originated from (using an lvalue-to-rvalue cast). And I don't think you should be trying to change the value of this original glvalue (I think this is what you mean by the "root" expression?):

*  I don't think there's a reliable way of finding this original glvalue. `TK_IgnoreUnlessSpelledInSource` looks like an attempt to do this that works in some cases, but I'm sure there are other cases where this doesn't work.

*  Indeed, there may not even be any glvalue that the prvalue originated from.

More generally, attempting to do this goes against the grain of value categories in C++. The whole idea of a prvalue is that it is a "pure value" with no storage location associated with it. Operations on prvalues are pure expression evaluation in the sense that this works in functional languages -- they have no way of changing anything that is stored in memory, i.e. they are free of side-effects.

So a check shouldn't be trying to go out and find the glvalue that a prvalue might have been cast from, and then changing the value associated with that glvalue's storage location -- because that would be doing something that isn't possible in C++. If we want to model the fact that we have learned new information about a value, that should be done by changing the flow condition, rather than changing the value itself.

Feel free to ask back if you're unsure how you would do this in the context of this check.

As an aside, I think making this change would also eliminate the restriction that the check currently doesn't deal correctly with references to pointers (as you note in the PR description).

Discookie replied:
By matching modeled pointer values instead, TK_IgnoreUnlessSpelledInSource is not needed anymore, as the framework does that work for me.

Discookie committed:
Review 1

* Add handling for `!= nullptr`, `!= ptr` operators
* Re-add TK_IgnoreUnlessSpelledInSource
* Refactor handling for null-check expressions (transfer instead of branchTransfer)
* Fix crash from `&(PrVar)` not always being PrVar
* Fix formatting and docs

Discookie committed:
Fix upstream fn signature change

EugeneZelenko commented on clang-tools-extra/docs/ReleaseNotes.rst:
```diff
@@ -110,6 +110,12 @@ New checks
+- New :doc:`bugprone-null-check-after-dereference
+  <clang-tidy/checks/bugprone/null-check-after-dereference>` check.
+
+  This check identifies redundant pointer null-checks, by finding cases where
```
Please omit `This check`.

EugeneZelenko commented on clang/lib/Analysis/FlowSensitive/Models/NullPointerAnalysisModel.cpp:
```diff
@@ -217,34 +231,72 @@ void matchDereferenceExpr(const Stmt *stmt,
-  Value *NewRootValue = Env.createValue(Var->getType());
+  // (bool)p or (p != nullptr)
+  bool IsNonnullOp = true;
+  if (auto *BinOp = dyn_cast<BinaryOperator>(NullCheck);
```
`
```suggestion
  if (const auto *BinOp = dyn_cast<BinaryOperator>(NullCheck);
```

Discookie committed:
Formatting

Discookie committed:
Aggressive Top skipping, first pass

Discookie committed:
Add support for function calls that take a pointer-of-pointer argument

Discookie committed:
Remove leftover logs

Discookie committed:
Fix crashes related to setting gl/prvalues mismatch

Discookie committed:
Remove reference to ControlFlowContext

Discookie committed:
Fix formatting

Discookie commented:
Should be ready for another round of review.

Simplified a lot of the logic, implemented most of the changes requested, added support for `p != nullptr`, and `void f(int **ptr)` invalidating the value of `*ptr`.

I've been experimenting with trying to remove `TK_IgnoreUnlessSpelledInSource`, without much luck. I need to be able to set a new `Value` to all of my pointers, and currently the framework doesn't let me set it in a persistent manner.
I'm still yet to refactor handling `Top`, currently it's used in early returns where possible, but I still need to think of a clear way to make central functions for it.

Discookie commented:
gentle ping @martinboehme @gribozavr

martinboehme commented:
> gentle ping @martinboehme @gribozavr

Sorry, I didn't notice your original comment that this is ready for review again.

I've added a few individual comments, but I haven't reviewed the whole PR in depth because I see that there are still quite a number of old comments that haven't been addressed. Please could you address these comments before the next review round?

Also:

> I'm still yet to refactor handling `Top`, currently it's used in early returns where possible, but I still need to think of a clear way to make central functions for it.

Can you address this too before the next review round? It's hard to review a PR that the author themself acknowledges isn't in a final state yet.

If you're not sure what the best way is to do certain things, happy to help with that of course -- but then we should address those specific topics first before continuing to review the whole PR.

Discookie committed:
Better top-bool- and generic-value handling

Discookie committed:
Remove TK_IgnoreUnlessSpelledInSource

Discookie committed:
Use new caller interface for runDataflowAnalysis

Discookie committed:
Merge branch 'main' into reverse-null

Discookie committed:
Merge branch 'main' into reverse-null

Discookie committed:
Merge branch 'main' into reverse-null

Discookie committed:
Simplify join logic, add some proper unit tests

Discookie committed:
Merge branch 'main' into reverse-null

Discookie commented on clang/include/clang/Analysis/FlowSensitive/Models/NullPointerAnalysisModel.h:
```diff
@@ -0,0 +1,112 @@
+  };
+
+  using ResultType =
+      std::pair<std::vector<SourceLocation>, std::vector<SourceLocation>>;
```
Change the value to be more self-explanatory as well.

Discookie reviewed (commented):
I'm still on this checker on and off, and there's been a bit of development since my last comments.
Added handling and early returns for Top values, fixed more crashes, and I'm currently in the process of making the logic simpler, eliminating unneeded solver invocations. Fixed all the review comments, except for the one related to `unpackPointerValue`, which I still need to figure out a good system for.

Discookie committed:
Minor fixes

denzor200 commented on clang-tools-extra/docs/clang-tidy/checks/bugprone/null-check-after-dereference.rst:
```diff
@@ -0,0 +1,163 @@
+       *p = 20; // false-positive: note tag not appended here
+     }
+
+     return ptr ? *ptr : 0;
```
`return p ? *p : 0;` ?

Discookie committed:
Fix docs

Discookie committed:
Simplify logic for compare and widen operations

Discookie committed:
Fix typo causing merging to fail

EugeneZelenko commented on clang-tools-extra/docs/clang-tidy/checks/bugprone/null-check-after-dereference.rst:
```diff
@@ -0,0 +1,163 @@
+
+The check currently supports null-checks on pointers that use
+``operator bool``, such as when being used as the condition
+for an ``if`` statement. It also supports comparisons such as ``!= nullptr``, and
```
Please follow 80 characters limit.

Discookie committed:
Merge branch 'main' into reverse-null

Discookie committed:
Fix formatting in docs

Discookie committed:
Fix code formatting

Discookie committed:
Merge branch 'main' into reverse-null

Discookie reviewed (commented):
I think this PR should do with another review - there's been significant refactors since the last one.

Discookie committed:
Merge branch 'main' into reverse-null

Discookie committed:
Lint and test fixes

Discookie committed:
Merge branch 'main' into reverse-null

Discookie committed:
Linter fixes

Discookie committed:
Linter fixes (again)
